packages/tracking/src/track.ts

```diff
@@ -0,0 +1,77 @@
+/* eslint-disable no-console */
+import type {
+  TrackingOptions,
+  EventDataTypes,
+  UserClickData,
+  UserVisitData,
+} from './types';
+
+export const createTracker = (apiBaseUrl: string, authToken: string, verbose = false) => {
+  const beacon = (endpoint: string, data: Record<string, string>) => {
+    const uri = new URL(endpoint, apiBaseUrl);
+    const searchParams = new URLSearchParams(uri.search);
+    searchParams.set('authentication_token', authToken);
+    uri.search = searchParams.toString();
+    const form = new FormData();
+    for (const [k, v] of Object.entries(data)) {
+      form.append(k, v.toString());
+    }
+
+    navigator.sendBeacon(uri.toString(), form);
+  };
+
+  const event = <
+    Category extends keyof EventDataTypes,
+    Event extends keyof EventDataTypes[Category],
+    Data extends EventDataTypes[Category][Event]
+  >(
+    category: Category,
+    event: Event,
+    userData: Data,
+    options: TrackingOptions = {}
+  ) => {
+    const properties = {
+      ...userData,
+      fullpath: window.location.pathname + window.location.search,
+      search: window.location.search,
+      path: window.location.pathname,
+      title: window.document.title,
+      url: window.location.href,
+    };
+
+    if (verbose) {
+      console.groupCollapsed(
+        `%cTracking Event Fired: ${category}:${event}`,
+        'color: #4b35ef; font-style: italic;'
+      );
+      console.log({
+        category,
+        event,
+        properties,
+      });
+      console.groupEnd();
+    }
+
+    beacon(`/analytics/${category}`, {
+      category,
+      event: event as string,
```

Comment:
Aha! I was wondering if you were going to hit this wall 😄  how irritating was this "event isn't assignable to string" error to get in dev?

```
Type 'Event' is not assignable to type 'string'.
  Type 'keyof EventDataTypes[Category]' is not assignable to type 'string'.
    Type 'string | number | symbol' is not assignable to type 'string'.
      Type 'number' is not assignable to type 'string'.ts(2322)
```

For visibility, the reason why this is coming up is that the `keyof` operator used above in `Event extends keyof EventDataTypes[Category]` returns a `string | number | symbol`. You can tell TypeScript you _only_ want the string keys of the object _(which is technically more correct & true)_:

```ts
Event extends string & keyof EventDataTypes[Category],
```

...and then you won't need the cast 😄 

```suggestion
      event,
```

🧙 ✨

Reply:
OMG.  Thank you.